systemd/datadog_checks/systemd/systemd.py

```diff
@@ -0,0 +1,155 @@
+# (C) Datadog, Inc. 2019
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from collections import defaultdict
+import pystemd
+from pystemd.systemd1 import Manager
+from pystemd.systemd1 import Unit
+
+from datadog_checks.base import AgentCheck, ConfigurationError, is_affirmative
+from datadog_checks.utils.subprocess_output import (
+    get_subprocess_output,
+    SubprocessOutputEmptyError,
+)
+
+
+class SystemdCheck(AgentCheck):
+
+    UNIT_STATUS_SC = 'systemd.unit.active'
+
+    def __init__(self, name, init_config, instances):
+        if instances is not None and len(instances) > 1:
+            raise ConfigurationError('Systemd check only supports one configured instance.')
+        super(SystemdCheck, self).__init__(name, init_config, instances)
+
+        instance = instances[0]
+        self.collect_all = is_affirmative(instance.get('collect_all_units', False))
+        self.units = instance.get('units', [])
+        # to store the state of a unit and compare it at the next run
+        self.unit_cache = defaultdict(dict)
+
+        # unit_cache = {
+        #    "units": {
+        #        "<unit_name>": "<unit_status>",
+        #        "cron.service": "inactive",
+        #        "ssh.service": "active"
+        #    }
+        # }
+
+    def check(self, instance):
+        if self.units:
+            for unit in self.units:
+                self.get_unit_state(unit)
+                self.get_number_processes(unit)
+        if self.collect_all:
+            # we display status for all units if no unit has been specified in the configuration file
+            self.get_active_inactive_units()
+
+        self.get_all_units(self.units)
+
+    def get_all_units(self, cached_units):
+        if not cached_units:
+            cached_units = self.unit_cache.get('units')
+
+        updated_units = self.get_listed_units()
+
+        # Initialize or update cache for this instance
+        self.unit_cache = {
+            'units': updated_units
+        }
+
+        return self.unit_cache, cached_units
+
+    def get_listed_units(self):
+        manager = Manager()
+        manager.load()
+        units = self.units
+
+        return {unit: self.get_state_single_unit(unit) for unit in units}
+
+    def update_unit_cache(self):
+        returned_cache = {}
+        returned_cache = self.get_listed_units()
+
+        return returned_cache
+
+    def get_active_inactive_units(self):
+        # returns the number of active and inactive units
+        manager = Manager()
+        manager.load()
+        list_units = manager.Manager.ListUnitFiles()
+
+        # remove units that have an @ symbol in their names - cannot seem to get unit info then - to investigate
```

Comment:
Just dropping a note that this is an important feature for us. Almost all of our systemd units are run using the pattern:
`service-component@0.service`
`service-component@1.service`
`service-component@2.service`

Where we need to run multiple copies of a service; for example when running multiple copies of an application's background threads.

`service-sidekiq-realtime@0.service`
`service-sidekiq-realtime@1.service`
`service-sidekiq-invoicing@0.service`

or running multiple web services as an upstream for nginx
`nginx.service`
`service.socket`
`service-webserver@0.service`
`service-webserver@1.service`

Reply:
Thanks for the note @wevanscfi. I just ran some more tests with 2 units that include an `@` symbol in their names, `serial-getty@ttyS0.service` and `ifup@enp0s3.service`. I did not run into any issue to get those unit states, so I must have made a mistake there. 
However, I'll do some additional testing to make sure we're properly handling this case.

Reply:
Investigated in https://github.com/facebookincubator/pystemd/issues/20. Unit names with `@` are supported.